File: task_server/api_testing/services/load_service_facts.py
"""Explicit operator-recorded service facts; never discover topology from prose."""
import copy
import hashlib
from ..executor import redact


COMPONENTS = {'database': 'database', 'downstream': 'downstream', 'gc_runtime': 'memory'}
BEHAVIORS = {'bounded_work_slots': 'application_pool', 'bounded_connection_slots': 'application_pool',
             'cpu_wall_time_work': 'cpu', 'shared_ttl_allocation': 'memory'}
STATES = {'present', 'absent', 'unknown'}
PARAMETERS = {
    'bounded_work_slots': {'limit'}, 'bounded_connection_slots': {'limit'},
    'cpu_wall_time_work': {'duration_ms', 'endpoints'},
    'shared_ttl_allocation': {'bytes', 'ttl_seconds', 'refresh_on_request', 'endpoints'},
}


def _require(condition, reason):
    if not condition:
        raise ValueError('服务能力事实：' + reason)


def normalize_service_facts(value):
    _require(isinstance(value, dict) and not set(value) - {'version', 'components', 'behaviors'}, '结构无效')
    _require(type(value.get('version')) is int and value['version'] == 1, '仅支持版本 1')
    result = {'version': 1, 'components': [], 'behaviors': []}
    for field, known, key in [('components', COMPONENTS, 'key'), ('behaviors', BEHAVIORS, 'kind')]:
        rows = value.get(field, [])
        _require(isinstance(rows, list) and len(rows) <= len(known), '条目数量无效')
        seen = set()
        for row in rows:
            _require(isinstance(row, dict), '条目必须为对象')
            name, state = row.get(key), row.get('state')
            _require(isinstance(name, str) and name in known and name not in seen, '类型无效或重复')
            _require(isinstance(state, str) and state in STATES, '状态无效')
            seen.add(name)
            params = PARAMETERS[name] if field == 'behaviors' and state == 'present' else set()
            _require(not set(row) - ({key, 'state', 'source'} | params) and params <= set(row), '行为参数与状态不匹配')
            item = {key: name, 'state': state}
            source = row.get('source')
            _require(state == 'unknown' or isinstance(source, dict), '已确认状态必须填写人工来源')
            if source is not None:
                _require(isinstance(source, dict) and not set(source) - {'kind', 'reference', 'recorded_by_operator'}, '来源结构无效')
                _require(source.get('kind') in ('operator_declaration', 'source_review', 'runtime_observation'), '来源类型无效')
                ref = source.get('reference')
                _require(isinstance(ref, str) and 0 < len(ref.strip()) <= 500, '来源引用必须填写且不超过 500 字符')
                item['source'] = {'kind': source['kind'], 'reference': redact(ref.strip()), 'recorded_by_operator': True}
            for param in params:
                v = row[param]
                if param == 'endpoints':
                    _require(isinstance(v, list) and 1 <= len(v) <= 20 and all(
                        isinstance(p, str) and 1 <= len(p) <= 500 and p.startswith('/') and not p.startswith('//')
                        and not any(c in p for c in ('?', '#', '\n', '\r')) for p in v), '接口范围必须为 1 至 20 条不带查询的相对路径')
                    item[param] = list(dict.fromkeys(v))
                elif param == 'refresh_on_request':
                    _require(type(v) is bool, '续期行为必须明确为是或否')
                    item[param] = v
                else:
                    maximum = {'limit': 1000000, 'duration_ms': 3600000, 'bytes': 1099511627776, 'ttl_seconds': 86400}[param]
                    _require(type(v) is int and 1 <= v <= maximum, '数值参数超出范围')
                    item[param] = v
            result[field].append(item)
    return result


def snapshot_service_facts(service_metadata, definition):
    """Freeze only declared facts for services/endpoints in this admitted scenario."""
    services = {}
    for step in definition.get('steps', []):
        request = step.get('request') or {}
        if not request:
            continue
        key = request.get('service', 'default')
        item = services.setdefault(key, {'step_ids': [], 'paths': []})
        item['step_ids'].append(step['id'])
        item['paths'].append(request.get('path'))
    for key, item in services.items():
        raw = (service_metadata.get(key) or {}).get('load_service_facts')
        normalized = normalize_service_facts(raw) if raw is not None else {'components': [], 'behaviors': []}
        item['components'] = copy.deepcopy(normalized['components'])
        item['behaviors'] = []
        for row in normalized['behaviors']:
            matching_steps = [step_id for step_id, path in zip(item['step_ids'], item['paths'])
                              if 'endpoints' not in row or path in row['endpoints']]
            if matching_steps:
                item['behaviors'].append({**copy.deepcopy(row), 'step_ids': matching_steps})
        del item['paths']
    return {'version': 1, 'services': services}


def fact_catalog(report):
    """A missing historical snapshot stays unknown; never look up live settings."""
    snapshot = ((report.get('evidence') or {}).get('environment_snapshot') or {}).get('service_facts') or {}
    services = snapshot.get('services') if snapshot.get('version') == 1 else None
    services = services if isinstance(services, dict) and services else {None: {}}
    result = []
    selected = list(services.items())[:30]
    if len(services) > 30:
        selected.append((None, {}))  # Truncated coverage cannot establish universal absence.
    for key, service in selected:
        suffix = hashlib.sha256(str(key).encode()).hexdigest()[:16]
        for field, known, name_key in [('components', COMPONENTS, 'key'), ('behaviors', BEHAVIORS, 'kind')]:
            rows = {r[name_key]: r for r in service.get(field, []) if isinstance(r, dict) and r.get(name_key) in known}
            for component, domain in known.items():
                row = rows.get(component, {})
                result.append({'evidence_id': f'fact.{suffix}.{component}', 'domain': domain, 'component': component,
                               'service_key': key, 'step_ids': list(row.get('step_ids', []) if field == 'behaviors' and 'endpoints' in row else service.get('step_ids', [])),
                               'state': row.get('state', 'unknown'), 'source': copy.deepcopy(row.get('source')),
                               'parameters': {k: copy.deepcopy(v) for k, v in row.items() if k not in {name_key, 'state', 'source', 'step_ids'}}})
    return result

File: api-testing-ui/src/utils/loadServiceFacts.ts
export type FactState = 'present' | 'absent' | 'unknown'
export type FactSource = {kind: string; reference: string; recorded_by_operator?: boolean}
export type ServiceFact = {key?: string; kind?: string; state: FactState; source?: FactSource; [key: string]: unknown}
export type ServiceFacts = {version: number; components: ServiceFact[]; behaviors: ServiceFact[]}
export const factDefinitions = [
  {key:'database', label:'数据库', group:'components'},
  {key:'downstream', label:'下游依赖', group:'components'},
  {key:'gc_runtime', label:'垃圾回收运行时', group:'components'},
  {key:'bounded_work_slots', label:'工作槽位上限', group:'behaviors'},
  {key:'bounded_connection_slots', label:'连接槽位上限', group:'behaviors'},
  {key:'cpu_wall_time_work', label:'固定墙钟时间 CPU 工作', group:'behaviors'},
  {key:'shared_ttl_allocation', label:'共享 TTL 内存分配', group:'behaviors'},
] as const
export const sourceLabels: Record<string,string> = {operator_declaration:'操作人声明',source_review:'代码核查记录',runtime_observation:'运行观察记录'}
export const stateLabels: Record<string,string> = {present:'存在',absent:'不存在',unknown:'未知'}
export function serviceFacts(metadata: Record<string,unknown>): ServiceFacts | undefined {
  return metadata.load_service_facts as ServiceFacts | undefined
}
export function serviceFactsIssue(metadata: Record<string,unknown>): string {
  const facts = serviceFacts(metadata)
  if (!facts) return ''
  if (facts.version !== 1 || !Array.isArray(facts.components) || !Array.isArray(facts.behaviors)) return '服务事实格式无法编辑，请核对记录版本'
  for (const row of [...facts.components,...facts.behaviors]) {
    const label = factDefinitions.find(item=>item.key===(row.key || row.kind))?.label || '服务事实'
    if ((row.state !== 'unknown' || row.source) && (!sourceLabels[row.source?.kind || ''] || !row.source?.reference.trim() || row.source.reference.length > 500)) return `${label}：请填写来源与引用（最多 500 字）`
    if (row.state !== 'present' || !row.kind) continue
    const ranges: Record<string,number> = row.kind.includes('slots') ? {limit:1000000} : row.kind === 'cpu_wall_time_work' ? {duration_ms:3600000} : {bytes:1099511627776,ttl_seconds:86400}
    if (Object.entries(ranges).some(([key,max])=>!Number.isSafeInteger(row[key]) || Number(row[key]) < 1 || Number(row[key]) > max)) return `${label}：请填写有效的正整数参数`
    if (row.kind === 'shared_ttl_allocation' && typeof row.refresh_on_request !== 'boolean') return `${label}：请选择请求时是否刷新 TTL`
    if (!row.kind.includes('slots') && (!Array.isArray(row.endpoints) || !row.endpoints.length || row.endpoints.length > 20 || row.endpoints.some(path=>typeof path !== 'string' || path.length > 500 || !path.startsWith('/') || path.startsWith('//') || /[?#\r\n]/.test(path)))) return `${label}：请填写 1–20 条接口路径，每行一条，以 / 开头`
  }
  return ''
}
